stonesoup/models/measurement/nonlinear.py

```diff
@@ -965,3 +967,192 @@ def rvs(self, num_samples=1, **kwargs) -> Union[StateVector, StateVectors]:
+            `False`, in which case no noise will be added and no binning takes place
+            if 'True', the output of :meth:`~.Model.rvs` is added and the
+            range and range rate are binned)
+
+        Returns
+        -------
+        :class:`numpy.ndarray` of shape (:py:attr:`~ndim_state`, 1)
+            The model function evaluated given the provided time interval.
+        """
+
+        out = super().function(state, noise, **kwargs)
+
+        if isinstance(noise, bool) or noise is None:
+            if noise:
+                out[0] = np.floor(out[0] / self.range_res) * self.range_res
+                out[1] = np.floor(
+                    out[1] / self.range_rate_res) * self.range_rate_res
+
+        return out
+
+    def _gaussian_integral(self, a, b, mean, cov):
+        # the cumlative probability ranging from a to b for a normal distribution
+        # \frac{1}{\sigma\sqrt{2\pi}}\int_{a}^{b}e^{-\frac{1}{2}\left(\frac{x-\mu}{\sigma}\right)^{2}}
+        return (multivariate_normal.cdf(a, mean=mean,
+                                        cov=cov) - multivariate_normal.cdf(
+            b, mean=mean, cov=cov))
+
+    def _binned_pdf(self, state_vector, mean, bin_size, cov):
+        # this function finds the probability density at state_vector1
+        # by dividing the intergral of a probability density function, over a bin, by the bin size
+        a = np.floor(state_vector / bin_size) * bin_size + bin_size
+        b = np.floor(state_vector / bin_size) * bin_size
+        return self._gaussian_integral(a, b, mean, cov)
+
+    def pdf(self, state1, state2, **kwargs):
+        r"""Model pdf/likelihood evaluation function
+
+        Evaluates the pdf/likelihood of ``state1``, given the state
+        ``state2`` which is passed to :meth:`function()`.
+
+        For the first 2 dimensions, this can be written as:
+
+        .. math::
+
+            p = p(y_t | x_t) = \mathcal{N}(y_t; x_t, Q)
+
+        where :math:`y_t` = ``state_vector1``, :math:`x_t` = ``state_vector2``,
+         :math:`Q` = :attr:`covar` and :math:`\mathcal{N}` is a normal distribution
+
+        The probability for the binned dimensions, the last 2, can be written as:
+
+        .. math::
+
+            p = P(a \leq \mathcal{N} \leq b)
+
+        In this equation a and b are the edges of the bin.
+
+        Parameters
+        ----------
+        state1 : :class:`~.State`
+        state2 : :class:`~.State`
+
+        Returns
+        -------
+        : :class:`~.Probability`
+            The likelihood of ``state1``, given ``state2``
+        """
+
+        # state_vector1 is in measurement space
+        # state_vector2 is in state_space
+        if ((state1.state_vector[0, 0] / self.range_res).is_integer()
+                and (state1.state_vector[
+                         1, 0] / self.range_rate_res).is_integer()):
+            mean_vector = self.function(state2, noise=False, **kwargs)
```

Comment:
Do I understand correctly that if the state has not been binned the PDF is zero. Is that correct? And should it be that way?

Reply:
I'm trying to check if the measured range and range rate is in one of the bins by seeing if it is a multiple of the bin size. let me know if there's a better way to do that.

Reply:
I have spotted an error. It's looking at the 2 angles. I'll change it to the range and range rate.

Reply:
So the check is: is state1.state_vector at a bin edge in both range and range rate. If not, then this is an impossibility and so the probability of it occurring is 0. Is that the logic?

Reply:
yes. I might be wrong but my understanding is pdf returns the likeliness of state1 if self.function was given state2. Does it seem right to you?

Reply:
I think you are right that the logic is correct then, but I had some issues with calculations when trying to work with delta-function-like PDFs. It means the area under the PDF is quite limited and, depending on how the function is used, it might never check that *specific* value. I think we would have a lot more confidence if there was a test/example that worked here.

Reply:
Just minor issue with the grammar on the documentation:
The covariances for radar **are** determined by different factors. The angle error
    is **affected** by the radar beam width. Range error is **affected** by the SNR and pulse bandwidth.
    The error for the range rate **is dependent** on the dwell time.
    The range and range rate are binned using

Reply:
Thank you for checking the grammar

Reply:
Thank you for checking the grammar.